### DISCUSSION TITLE
How to Restore the Original Commit History?

### DISCUSSION THREAD

**Discussion by sanjeevmansotra1** [community/community#106718] [Repositories]
### Select Topic Area

Question

### Body

Hello my name is Sanjeev Mansotra. One of your team members accidentally force-pushed to the master branch, overwriting some important commits. As a result, the commit history is now different from the remote repository. How to revert the force-push and restore the original commit history?
Explain the step-by-step process you would follow to achieve this, considering the need to communicate effectively with your team and maintain a clean version history. 
Waiting for your response! 
Thanks :)

**Comment by subidit** [comment]
- Switch to your main branch with `git checkout main`
- Update your local main branch with `git pull`. This is so we have the latest HEAD of main available for the rebase.
- Switch back to your branch foo with `git checkout foo`
- Use `git rebase main`, this will complete the rebase, replaying your commits on top of the HEAD of main.

For a full explanation check this [link](https://www.freecodecamp.org/news/how-to-use-git-rebase/).

**Comment by srivarshithdaladuli** [comment]
1. Communicate to team to stop new commits:

```
@team Hold off on new commits to master. Accidental force-push overwritten history which I'm going to restore.
```

2. Locally revert force-pushed commits: 

```
git fetch origin
git reset --hard origin/master
```

3. Restore original commit history:

```
# Checkout old master 
git checkout master-before-force-push

# Rebase to get back old commits
git rebase -i master

# Force push to restore history 
git push --force origin master
```

4. Inform team master is restored:

```  
@team Original commit history is now restored on master. Please create a new branch for any risky changes to avoid disrupting master again.
```

Let me know if you need any clarification or have additional questions!

**Comment by Krishna-69** [comment]
You can use the git revert command to return your repository's files to a previous state without rewriting the commit history. This is done by creating new commits that do the opposite of existing commits, i.e. removing lines and files that were added and adding lines and files that were removed.

**Comment by srivarshithdaladuli** [comment]
git revert commit

**Comment by queenofcorgis** [comment]
Thanks for posting in the GitHub Community, @sanjeevmansotra1 ! We're happy you're here. You are more likely to get a useful response if you are posting your question in the applicable category, the Discussions category is solely related to conversations around the GitHub product Discussions. I've gone ahead and moved it for you. Good luck!
